### ISSUE TITLE
Incremental compilation fails with "Number of loaded files in snapshots differs"

### ISSUE THREAD

**Issue by graemerocher** [google/ksp#1280]
We are building support for KSP in Micronaut (a popular Java/Kotlin framework). The initial implementation is done but it fails with KSP incremental compilation enabled with an error "Number of loaded files in snapshots differs" on the second compile:

```
./gradlew test
> Task :kspKotlin FAILED

FAILURE: Build completed with 2 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':kspKotlin'.
> Number of loaded files in snapshots differs. Reported changed files: []
```

To reproduce download this example project and run `./gradlew cTK` twice.

[ksp-test2.zip](https://github.com/google/ksp/files/10480874/ksp-test2.zip)

**Comment by SimonMarquis** [comment]
👋 This issue popped up again since [1.8.20-Beta-1.0.9](https://github.com/google/ksp/releases/tag/1.8.20-Beta-1.0.9).

Here is an example with [1.8.20-RC-1.0.9](https://github.com/google/ksp/releases/tag/1.8.20-RC-1.0.9) on the [android/nowinandroid](https://github.com/android/nowinandroid) project:
https://github.com/SimonMarquis/nowinandroid/actions/runs/4409825508/jobs/7726595603#step:9:1881

Should we create a new issue?

**Comment by neetopia** [comment]
@SimonMarquis I can't reproduce the bug consistently with a locally checked out repository, can you verify on your side with a snapshot build `1.8.20-RC-1.0.10-SNAPSHOT` from repository `https://oss.sonatype.org/content/repositories/snapshots/`?

**Comment by neetopia** [comment]
for context, the snapshot contains fix from #1350 which I anticipate to fix

**Comment by SimonMarquis** [comment]
Sorry for not being able to test this, I did not found enough time to test it yet.
Though it still happens on Kotlin 1.8.20 and KSP 1.8.20-1.0.10.
https://github.com/SimonMarquis/nowinandroid/actions/runs/4600305482/jobs/8126733329#step:9:2079

It might very well be an issue with the non-statble (non-compatible?) Compose compiler 1.4.4-dev-k1.8.20-RC-88d9f3a8232.

**Comment by ZacSweers** [comment]
@neetopia did #1350 make it into 1.8.20-1.0.10? I'm still reproducing this in that release now in https://github.com/slackhq/circuit/pull/454.

100% reproducible with the below steps
1. run `./gradlew check`
2. make a small change to a source test file, such as adding a println in `PetListSnapshotTest.kt`
3. run `./gradlew check` again
4. 💥

**Comment by neetopia** [comment]
yes #1350 was included in the release, but does not seem to fix the issue.

**Comment by neetopia** [comment]
I was able to reproduce @ZacSweers 's repro, but after I deleted both `$HOME/.gradle/caches` and project's `.gradle/caches`,  and tries to do the reproduce steps again, it seems not reproducing anymore..

**Comment by Enaium** [comment]
I also encountered this problem when using 1.8.20-1.0.10, Need to run `gradle clean` before run the project each time

**Comment by neetopia** [comment]
@Enaium can you give me a reproduce project? I am not able to reproduce @ZacSweers 's sample anymore after manually removing home caches and project configuration caches. Right now I am not sure if that has anything to do with remote cache.

**Comment by ZacSweers** [comment]
I'm able to reproduce it 100%. Maybe try building once with 1.8.10 and then 1.8.20?

**Comment by neetopia** [comment]
```
7168  gco main
 7169  less gradle.properties
 7170  less gradle/libs.versions.toml
 7171  ./gradlew check
 7172  git log
 7173  gco z/kotlin1820
 7174  ./gradlew check
 7175  vi samples/star/src/test/kotlin/com/slack/circuit/star/petlist/PetListSnapshotTest.kt
 7176  ./gradlew check
```

yeah that was what I did try to reproduce. I was able to reproduce 100% as well before deleting both local caches. I have another laptop not with me that can still reproduce 100% as I did not try removing caches on that laptop.

**Comment by Enaium** [comment]
> @Enaium can you give me a reproduce project? I am not able to reproduce @ZacSweers 's sample anymore after manually removing home caches and project configuration caches. Right now I am not sure if that has anything to do with remote cache.
https://github.com/Enaium/jimmer-sql-kt-vertx

**Comment by svenjacobs** [comment]
I'm facing the same issue since upgrading to Kotlin `1.8.20` and KSP `1.8.20-1.0.10`. Interestingly I cannot reproduce this on my local machine but 100% in a GitHub Actions workflow, although the workspace is absolutely clean with no caches (I disabled the cache action to ensure this).

**Comment by G00fY2** [comment]
Same issue here. We run our Jenkins CI builds without any caching in place, so 100% clean build each time. After updating to Kotlin 1.8.20, KSP 1.8.20-1.0.10 and Compose compiler 1.4.5 we also see this build error.

**Comment by neetopia** [comment]
Hi all, thanks for the reports, I've pushed a fix in #1363 and it is now available as `1.8.20-1.0.11-SNAPSHOT` in snapshot repository, you can add 
```
  maven {
    url = uri("https://oss.sonatype.org/content/repositories/snapshots")
  }
```
in your project and try out this snapshot to see if it resolves your issue, if everything looks good I will do a hot fix release soon to address this issue.

**Comment by jeffdcamp** [comment]
So glad I found this issue.... we couldn't run after updating to Kotlin 1.8.20, KSP 1.8.20-1.0.10 and Compose compiler 1.4.5.

We just tried the 1.8.20-1.0.11-SNAPSHOT and all is working for us!  (hope 1.8.20-1.0.11 releases soon :) )

**Comment by hansenji** [comment]
This fixed it for me as well in my CI/CD running Unit Tests

**Comment by masztalski** [comment]
Seems to work for me also. Waiting for final build

**Comment by neetopia** [comment]
Cool, hot-fix release should be available on Monday.

**Comment by svenjacobs** [comment]
I can also confirm that `1.8.20-1.0.11-SNAPSHOT` fixes the problem 🎉  Thank you!

**Comment by neetopia** [comment]
hot fix is released as 1.8.20-1.0.11, closing this issue.

**Comment by rschattauer** [comment]
@neetopia I am still facing this issue with `1.8.22-1.0.11`

**Comment by neetopia** [comment]
@rschattauer  Can you please give me a reproduce repo for this issue in 1.8.22-1.0.11? Thanks.

**Comment by pengdev** [comment]
Hey @neetopia I'm facing the same issue with [1.7.20-1.0.8](https://github.com/google/ksp/releases/tag/1.7.20-1.0.8) release, as we can not bump kotlin version in our project, is it possible to port the fix to the 1.7.20 kotlin release? Or could you suggest any workarounds that we can apply to our project to bypass this issue? Thanks in advance!

**Comment by neetopia** [comment]
We can't do a back port release, would disable incremental to mitigate this issue work for you?

**Comment by pengdev** [comment]
@neetopia Thanks for the suggestion, I did try to disable incremental and it did help mitigate this issue. I think we are ok with the workaround for now 👍

**Comment by speedclaud** [comment]
@liujigang Yes, I'm having the same issue with these versions after bumping: 

```
com.google.devtools.ksp: 2.1.20-1.0.32
com.android.tools.build:gradle:8.9.1
gradle:8.11.1
```

**Comment by Micoder-dev** [comment]
Same issue popped up again.
```id("com.google.devtools.ksp") version "2.1.20-1.0.32" apply false```

**Comment by luboganev** [comment]
I'm also experiencing the same issue again in Android Studio stable 
- Android Studio Meerkat | 2024.3.1 Patch 1.
- Kotlin 2.1.20
- KSP 2.1.20-1.0.32
- AGP 8.9.1
- gradle-8.13
- openjdk 17.0.14 2025-01-21
- Mac Apple M4 Pro
- macOS 15.4 (24E248)

The only current "fix" for me is downgrade:
- Downgrade to KSP 2.1.10-1.0.30
- Downgrade to Kotlin 2.1.10

**Comment by Micoder-dev** [comment]
@luboganev Yeah, It worked Thanks..!

**Comment by marianeum** [comment]
Is anybody able to create a project to reproduce the issue? We're affected as well, but I couldn't reproduce it on a small sample project. It might be worth creating a new issue with that as I'm not sure this is monitored anymore.

**Comment by luboganev** [comment]
@marianeum I'm trying to do it now

**Comment by luboganev** [comment]
@marianeum 

I have some updates. I have started as well a fresh project, but couldn't reproduce the issue. Than started adding more and more stuff to it until it failed.

In my actual project I use ksp for everything besides Retrofit (since AFAIK Retrofit doesn't have KSP yet). At the moment I have added Retrofit and kapt, I have started observing the issue. So it's something that goes wrong between Kapt and ksp. Here is the repository to reproduce it. If you remove the kapt stuff everything is fine, but if you add it, then 2 consecutive builds always fail if no clean is done between:

https://github.com/luboganev/kspandkaptissue

@ting-yuan , what would you recommend we do from here? Thanks a lot!

**Comment by ting-yuan** [comment]
While I'm still investigating this issue, it seems that it only happens on KSP1 but not KSP2. Could you [try KSP2](https://github.com/google/ksp/blob/main/docs/ksp2.md#using-ksp-2-in-gradle) as a workaround and let me know how it works?

**Comment by luboganev** [comment]
@ting-yuan 

I can confirm that the proposed workaround resolves the issue. I have implemented this in the following branch of the sample project:
https://github.com/luboganev/kspandkaptissue/tree/fix/enable-ksp2-workaround

Thank you very much investigation and the support!

**Comment by luboganev** [comment]
@ting-yuan , unfortunately, I was too quick to celebrate. Activating ksp2 now is suffering from this issue:
https://github.com/google/ksp/issues/2177

**Comment by consp1racy** [comment]
Don't close this. Ksp2 is a no go, it's painfully slow.

**Comment by G00fY2** [comment]
KSP2 is not supported by dagger/hilt yet. So not an option for now: https://github.com/google/dagger/issues/2349#issuecomment-2725553921

**Comment by luboganev** [comment]
Good day everyone!

Small update from my side. The latest release https://github.com/google/ksp/releases/tag/2.1.20-2.0.0 has the Kapt compilation fix in it, so it resolves the issue in the sample project.
